### PR TITLE
KAFKA-5602: ducker-ak: support --custom-ducktape

### PR DESCRIPTION
**Pull request by cmccabe**
Support a --custom-ducktape flag which allows developers to install
their own versions of ducktape into Docker images.  This is helpful for
ducktape development.

### CODE REVIEW

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6124/
Test PASSed (JDK 7 and Scala 2.11).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6109/
Test PASSed (JDK 8 and Scala 2.12).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6296/
Test PASSed (JDK 8 and Scala 2.12).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6313/
Test PASSed (JDK 7 and Scala 2.11).

File: tests/docker/ducker-ak [hunk 1]
```diff
@@ -58,10 +58,14 @@ help|-h|--help
     Display this help message
 
 up [-n|--num-nodes NUM_NODES] [-f|--force] [docker-image]
+        [-E|--custom-ducktape DIR]
```
Comment (ijuma) [inline tests/docker/ducker-ak]: How was the `E` chosen?

File: tests/docker/ducker-ak [hunk 2]
```diff
@@ -217,10 +221,40 @@ $((${duration} % 60))s.  See ${build_log} for details."
...
+        -d -t --net=host -h "${node}" --network ducknet \
+        --memory=${docker_run_memory_limit} --memory-swappiness=1 \
+        -v "${kafka_dir}:/opt/kafka-dev" --name "${node}" -- "${image_name}"
+}
+
+setup_custom_ducktape() {
+    local custom_ducktape="${1}"
+    local image_name="${2}"
+
+    [[ -f "${custom_ducktape}/README.md" ]] || \
```
Comment (ijuma) [inline tests/docker/ducker-ak]: Isn't there something more unique to ducktape that we can check?

File: tests/docker/ducker-ak [hunk 3]
```diff
@@ -217,10 +221,40 @@ $((${duration} % 60))s.  See ${build_log} for details."
...
+        die "You must supply a valid ducktape directory to --custom-ducktape"
+    docker_run ducker01 "${image_name}"
+    local running_container="$(docker ps -f=network=ducknet -q)"
+    must_do 1 1 docker cp "${custom_ducktape}" "${running_container}:/opt/ducktape"
+    docker exec --user=root ducker01 bash -c 'set -x && cd /opt/ducktape && python ./setup.py develop && cd /opt/kafka-dev/tests && python ./setup.py develop'
+    [[ $? -ne 0 ]] && die "failed to install the new ducktape."
+    must_do 1 1 docker commit ducker01 "${image_name}"
+    must_do 1 0 docker kill "${running_container}"
+    must_do 1 0 docker rm ducker01
```
Comment (ijuma) [inline tests/docker/ducker-ak]: Could we not use something a bit more descriptive for the must_do parameters? It's a little hard to read at the moment.

**Review by ijuma — COMMENTED** [review]
Thanks for the PR. This is very useful. Left a few minor comments. cc @ewencp who knows this code better than I do.

File: tests/docker/ducker-ak [hunk 4]
```diff
@@ -58,10 +58,14 @@ help|-h|--help
     Display this help message
 
 up [-n|--num-nodes NUM_NODES] [-f|--force] [docker-image]
+        [-E|--custom-ducktape DIR]
     Bring up a cluster with the specified amount of nodes (defaults to ${default_num_nodes}).
     The docker image name defaults to ${default_image_name}.  If --force is specified, we will
     attempt to bring up an image even some parameters are not valid.
 
+    If --custom-ducktape is specified, we will install the provided custom
```
Comment (ewencp) [inline tests/docker/ducker-ak:66]: Might want to clarify that this is the ducktape repo, not, e.g., the actual installed module.

File: tests/docker/ducker-ak (hunk 1 above)
Comment (cmccabe) [reply tests/docker/ducker-ak]: I changed it to -C to be a little more intuitive.

File: tests/docker/ducker-ak (hunk 2 above)
Comment (cmccabe) [reply tests/docker/ducker-ak]: Changed it to look for `${custom_ducktape}/ducktape/__init__.py`

File: tests/docker/ducker-ak (hunk 3 above)
Comment (cmccabe) [reply tests/docker/ducker-ak]: OK.  We can have flags.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6590/
Test PASSed (JDK 7 and Scala 2.11).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6575/
Test PASSed (JDK 8 and Scala 2.12).

**Review by ijuma — APPROVED** [review]
Thanks for the updates, LGTM. I'll merge to trunk. If @ewencp has additional comments, we can address in a follow-up.

**Pull request closed without merging**